src/main/java/com/openlattice/chronicle/services/message/MessageService.kt

```diff
@@ -0,0 +1,99 @@
+    }
+
+    override fun sendMessages(organizationId: UUID, messageDetailsList: List<MessageDetails>) {
+        val outcomes = twilioManager.sendMessages(messageDetailsList)
+        for(outcome in outcomes) {
+            try {
+                recordMessageSent(organizationId, outcome)
+                logger.info("Recorded notification {} sent to participant {}", outcome.sid, outcome.participantId)
+            } catch (e: ExecutionException) {
+                logger.error("Unable to record notification sent for SID {}", outcome.sid, e)
+            }
+        }
+    }
+
+    @Throws(ExecutionException::class)
+    fun recordMessageSent(organizationId: UUID?, messageOutcome: MessageOutcome) {
+        val apiClient = apiCacheManager.prodApiClientCache[ApiClient::class.java]
+        val dataApi = apiClient.dataApi
+
+        // entity set ids
+        val participantES = ChronicleServerUtil.getParticipantEntitySetName(messageOutcome.studyId)
+        val coreAppConfig = entitySetIdsManager
+            .getChronicleAppConfig(organizationId)
+        val legacyAppConfig = entitySetIdsManager
+            .getLegacyChronicleAppConfig(participantES)
+
+        val messageESID = coreAppConfig.messagesEntitySetId
+        val sentToESID = legacyAppConfig.sentToEntitySetId
+        val participantsESID = coreAppConfig.participantEntitySetId
+
+        val participantEKID =
+            enrollmentManager.getParticipantEntityKeyId(organizationId, messageOutcome.studyId, messageOutcome.participantId)
+
+        val entities: ListMultimap<UUID, Map<UUID, Set<Any>>> = ArrayListMultimap.create()
+        val associations: ListMultimap<UUID, DataAssociation> = ArrayListMultimap.create()
+
+        val messageEntity = mapOf(
+            edmCacheManager.getPropertyTypeId(EdmConstants.TYPE_FQN) to ImmutableSet.of<Any>(messageOutcome.messageType),
+            edmCacheManager.getPropertyTypeId(EdmConstants.OL_ID_FQN) to ImmutableSet.of<Any>(messageOutcome.sid),
+            edmCacheManager.getPropertyTypeId(EdmConstants.DATE_TIME_FQN) to ImmutableSet.of<Any>(messageOutcome.dateTime),
+            edmCacheManager.getPropertyTypeId(EdmConstants.DELIVERED_FQN) to ImmutableSet.of<Any>(messageOutcome.isSuccess)
+        )
+        entities.put(messageESID, messageEntity)
+
+        val sentToEntity = mapOf(
+            edmCacheManager.getPropertyTypeId(EdmConstants.DATE_TIME_FQN) to ImmutableSet.of<Any>(messageOutcome.dateTime)
+        )
+
+        associations.put(
+            sentToESID,
+            DataAssociation(
+                messageESID,
+                Optional.of(0),
+                Optional.empty(),
+                participantsESID,
+                Optional.empty(),
+                Optional.of(participantEKID),
+                sentToEntity
+            )
+        )
+        val dataGraph = DataGraph(entities, associations)
+        dataApi.createEntityAndAssociationData(dataGraph)
```

Comment:
Since you are calling `recordMessageSent` in a for loop, you calling `dataApi.createEntityAndAssociationData` once per message, which is not ideal. You should be able to make a single call to `dataApi.createEntityAndAssociationData` with all the data; please refactor.